### PR TITLE
feat: Enable StateClient usage under docker plugins.

### PR DESCRIPTION
**Pull request by marianogappa**
`StateClient`s currently don't work in docker-based plugins because they cannot connect to destination sockets.

This CLI change detects that a source has:
- the `Docker` registry setting
- the `backend_options` set

And in this case it makes the following config changes:
- Forces destinations to use TCP instead of sockets
- Adds the `host.docker.internal` extra host to the docker create configuration
- Sniffs the source's variables and replaces instances of localhost with this special hostname

As a result, normal syncs continue to work exactly the same as before, but syncs that suffered from this config problem now work seamlessly with the new settings:

Screenshot 2024-08-08 at 16 33 32
Screenshot 2024-08-08 at 16 36 32

(Note that I'm using an unreleased pb-go replace)

### CODE REVIEW

File: cli/cmd/sync.go [hunk 1]
```diff
@@ -195,6 +200,9 @@ func sync(cmd *cobra.Command, args []string) error {
 			managedplugin.WithTeamName(teamName),
 			managedplugin.WithLicenseFile(licenseFile),
 		}
+		if isThereADockerSourcePluginWithBackend {
```
Comment (erezrokah) [inline cli/cmd/sync.go:203]: Shouldn't we do this only for destinations that are used as state?

Comment (erezrokah) [reply cli/cmd/sync.go:203]: Also a follow PR can be to update the logic to consider if the OS is Windows per https://github.com/cloudquery/cloudquery/issues/18039

Comment (marianogappa) [reply cli/cmd/sync.go]: I made the following changes:
- UseTCP is more strict; only enabling it when the destination is used as state
- I propagated the state that a source `shouldReplaceLocalhost` all the way to `ReplaceVariables`. Before, it was replacing if a source had a `RegistryDocker`, which is not ideal because it doesn't necessarily need to use state. I don't think it's a bug because if it doesn't use a backend it probably doesn't need any replacing, but it's more correct now.
- Added tests where possible.

Regarding the Windows edge case, I'd fix it separately (in another PR, as you say)

**Review by erezrokah — APPROVED** [review]
